homeassistant/components/shelly_mqtt/switch.py

```diff
@@ -0,0 +1,167 @@
+"""Representation of Shelly MQTT switches."""
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.const import CONF_DEVICE_ID
+from homeassistant.core import callback
+
+from .const import (
+    COMMAND_SUFFIX,
+    CONF_MODEL,
+    CONF_TOPIC,
+    DOMAIN,
+    MODEL_SWITCHES,
+    MODEL_TITLE,
+    MODELS,
+)
+
+ON = "on"
+OFF = "off"
+TOGGLE = "toggle"
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Shelly MQTT switch(es) from config entry."""
+
+    device_id = config_entry.data[CONF_DEVICE_ID]
+    topic = config_entry.data[CONF_TOPIC]
+    model = config_entry.data[CONF_MODEL]
+    title = MODELS[model][MODEL_TITLE]
+    switches_count = MODELS[model][MODEL_SWITCHES]
+    switches = []
+    for i in range(switches_count):
+        switch_topic = topic + f"relay/{i}"
+        if i == 0:
+            unique_id = device_id
+            via_device = None
+        else:
+            unique_id = f"{device_id}_{i+1}"
+            via_device = device_id
+        switches.append(
+            ShellyMQTTSwitch(
+                hass.components.mqtt, unique_id, switch_topic, title, via_device
+            )
+        )
+
+    async_add_entities(switches)
+
+
+class ShellyMQTTSwitch(SwitchEntity):
+    """Representation of a Shelly MQTT switch."""
+
+    def __init__(self, mqtt, unique_id, topic, title, via_device):
+        """Initialize the MQTT switch."""
+        self._mqtt = mqtt
+        self._unique_id = unique_id
+        self._via_device = via_device
+        self._topic = topic
+        self._title = title
+        self._power = None
+        self._energy = None
+        self._unsubscribe = None
+        self._state = None
+
+    async def async_added_to_hass(self):
+        """Subscribe to MQTT events."""
+        await super().async_added_to_hass()
+        await self._subscribe_topics()
+
+    async def _subscribe_topics(self):
+        """(Re)Subscribe to topics."""
+
+        @callback
+        def state_message_received(msg):
+            """Handle new MQTT state messages."""
+            if msg.topic == self._topic:
```

Comment:
The topic parsing and publish and subscribe handling should be part of a 3rd party library published on pypi. See the mysensors and ozw integrations for examples.

Reply:
Might be a great opportunity to reuse the OZW MQTT base classes https://github.com/cgarwood/python-openzwave-mqtt/blob/master/openzwavemqtt/base.py

Reply:
@balloob I thought that importing from other integrations is not allowed. Is it?
If it is, I'd like to simply use the MQTT entity classes. I can just initialize them with the right topics, payloads, and device info.

Reply:
We don't want that, because we wouldn't want to end up having to make/not make changes to MQTT because Shelly MQTT relies on it.

Reply:
My intention is to use it just like you would from configuration.yaml, but with the added benefit of (manual) discovery, automatic setup of all the entities for a given device, and device and entity registry support. Any breaking change will most likely also break the regular configuration.

Reply:
Could we use some sort of "discovery mechanism" by which this component would forward setup of the MQTT entities to the MQTT integration ?

This mechanism could become quite generic (core) and be used by other integration (I have an idea for the utility_meter based on this)

Reply:
That's the approach that @bieniu is taking, but as an external script.
Maybe he has an idea how to bring it into HA as an integration.

Reply:
I thought that the MQTT based component would be the simplest and most reliable solution. The external script is very difficult to maintain due to the `python_script` limitation.

Reply:
I guess the question is can we use the same logic as the script, but in a component. So there's a config flow, but instead of directly creating devices, it'll feed the existing MQTT entities.
One way is to directly create MQTT entities, but @balloob is opposed to the idea.
Another is to instead send the discovery messages over MQTT, but I believe that would sever the connection between the config entry and the actual devices and entities.
We can also consider creating a higher level way to create MQTT entities, that does not rely on their internals. Imagine an API where you can create an MQTT switch/sensor/etc..., that all the integrations can use.

Reply:
MQTT discovery based on topics is out of scope for this PR, but is definitely an option.

Still, this integration would need to handle MQTT themselves instead of relying on the MQTT integration for anything but subscribing to topics. Have a look at the OZW base classes, they are perfect for this.

Reply:
@balloob I can see the appeal, but I think this might be overkill given how simple the Shelly MQTT protocol is. It'll end up creating a very complicated structure, with a tight dependency on a package, just to create a simplified version of the existing MQTT entities.
I think I'd rather go with topic based discovery - this should really simplify everything. The only open issue is the ability to tie the created devices and entities to the config entry. 
If I create a hub device, and then use `via_device` in the discovery info, would that work?

Reply:
Tried a little experiment (code below). The result was that the "hub" device was correctly attributed to the Shelly MQTT integration, but the "child" devices were attributed to the MQTT integration, and I could not pass the config entry id as part of the device info since the schema validation prevented it.


```
discovery_info = {
    "command_topic": f"{switch_topic}/{COMMAND_SUFFIX}",
    "device": {
        # "config_entry_id": config_entry.entry_id,
        "identifiers": [f"({DOMAIN}, {unique_id})"],
        "manufacturer": "Shelly",
        "model": title,
        "name": unique_id,
        "via_device": f"({DOMAIN}, {device_id})"

    },
    "name": unique_id,
    "payload_off": OFF,
    "payload_on": ON,
    "state_topic": switch_topic,
    "unique_id": unique_id
}

device_registry = await dr.async_get_registry(hass)

device_registry.async_get_or_create(
    config_entry_id=config_entry.entry_id,
    identifiers={(DOMAIN, device_id)},
    manufacturer="Shelly",
    name=device_id,
    model=title,
)

hass.components.mqtt.async_publish(f"homeassistant/switch/{unique_id}/config", json.dumps(discovery_info))
```

Reply:
Every integration stores the protocol in PyPI package. MQTT-based integrations will not be treated any differently.

Reply:
Understood. My point is not that there shouldn't be a package, but that the specific proposal of a highly abstracted package, coupled with the fact that I'm trying to recreate existing functionality, might suggest that a different solution might be better.
That's why I'm trying to go with the discovery topic approach (which, of course, will rely on a PyPI package for identifying devices, getting the right topics and payloads, etc...)

Reply:
@balloob before I invest more type in the discovery topic based approach, can you confirm it's an accepted one?
Also, are you open to core changes to allow control over the config entry id in such a case?

Reply:
I'd say it's not acceptable. We want the devices and entities to be created by the shelly mqtt integration and not the mqtt integration. The mqtt integration should just provide the transport for the communication.

Reply:
I would say MQTT is a very special case, it's a BUS it's not really an integration, we have various integrations that to not fall into HA ecosystem that piggyback on MQTT, shelly would be just another case except this one time it lives in the HA ecosystem.

That being said, I still thing a better option would be to use HA discovery mechanism and not MQTT discovery mechanism...

Reply:
This is exactly what we don't want. The MQTT integration has grown way too complex due to all the piggybacking. We shouldn't expand that.

Reply:
I can see the concern about the MQTT integration bloat. However, as we get more MQTT based integrations, we're seeing a lot of repeated code.
I think we can differentiate between the MQTT infrastructure and the MQTT integration. The integration does indeed cover a lot - optimistic updates, availability, json parsing, and discovery to name a few. But it's also a cliff - if you need some custom behavior (e.g. custom discovery), you're off the cliff and have to start from the ground up, based on the low level infrastructure (basically subscribe and publish).
What if we increased the scope of the infrastructure so that it allows integrations to take advantage of all of these higher level building blocks? That will also allow us to simplify the MQTT integration, as it will basically be just another integration built on top of this infrastructure.

Reply:
That's a topic for the architecture repo and not this PR. At least I won't discuss our architecture here. I can only tell you what our current standard is what changes are needed in this PR based on that.

Reply:
Makes sense:
https://github.com/home-assistant/architecture/issues/409